### ISSUE TITLE
 Real-time Search Suggestion Feature for Project Search Engine

### ISSUE THREAD

**Issue by nitinawari** [OWASP/Nest#148]
**Is your feature request related to a problem? Please describe.**
Currently, users have to manually type and search for items without any suggestions, which can lead to inefficiencies, especially when searching for common terms. It frustrates users when they have to guess or type the exact term to find what they are looking for.

**Describe the solution you'd like**
I want to implement a real-time search suggestion feature with the following behavior:

1 . As the user types in the search box, relevant tags from a predefined tags  are displayed in a dropdown.
   - For example: Typing "ja" should suggest "Java," "JavaScript," and "Java-agent."
The dropdown should be dynamically updated as the user continues typing.
Users can:
 - Click on a suggestion to select it.
 - Use arrow keys to navigate through suggestions and press Enter to select.
Upon selecting a tag, the search will find and retrieve folders linked to that tag.

**Describe alternatives you've considered**
1 . Static search results: 
Users manually type and submit their search queries, but this lacks real-time guidance and slows down the process.
2. Auto-complete with no dropdown:
Directly fills the search input but doesn’t provide a list of multiple matching tags, limiting user flexibility.

**Additional context**
  Attached below is an example image demonstrating the basic idea of the search suggestion feature: 
**Note**: The UI in the image is a rough draft; it will be refined during implementation.
![owasp1](https://github.com/user-attachments/assets/0f3dc89b-6c73-4d50-a6a9-6da52feaae53)
![owasp2](https://github.com/user-attachments/assets/208dc670-5352-4152-badb-ce7cc602dc71)

**Comment by nitinawari** [comment]
hey @arkid15r
What do you think of this feature? If you approve, I’ll start working on it as soon as possible. Please assign it to me. Happy coding!

**Comment by arkid15r** [comment]
> hey @arkid15r What do you think of this feature? If you approve, I’ll start working on it as soon as possible. Please assign it to me. Happy coding!

Hi @nitinawari 
thanks for the suggestion. I believe it may be a very valuable UX improvement if implemented correctly. How are you going to create the set of relevant auto-completion suggestions phrases though?

**Comment by arkid15r** [comment]
> Hi @arkid15r,
> 
> Thanks for the feedback! I’m currently figuring out the approach for the suggestions. I’ll update you here or on Slack once I have a solution.

Sounds good. I asked because I had looked into this some time ago and found out quite interesting way of how it could be done with almost no effort (I'm not sure about the quality of the suggestions). So I'm interested to see what would be your approad for tackling this.

**Comment by Rajgupta36** [comment]
@arkid15r Sir, I also have an idea of using manual API calls with debouncing to make the search more efficient. By implementing debouncing, we can reduce the number of API requests made while the user is typing, which would help avoid unnecessary calls and improve performance. This approach ensures that the search suggestions are updated in real-time without overwhelming the system with too many requests.

**Comment by arkid15r** [comment]
> @arkid15r Sir, I also have an idea of using manual API calls with debouncing to make the search more efficient. By implementing debouncing, we can reduce the number of API requests made while the user is typing, which would help avoid unnecessary calls and improve performance. This approach ensures that the search suggestions are updated in real-time without overwhelming the system with too many requests.

It seems your comment is more relevant to this issue -- https://github.com/OWASP/Nest/pull/145. And this technique is already in place for the search component. However, I'm not sure if @Yashgupta9330 still works on the PR.

**Comment by nitinawari** [comment]
@arkid15r Yes, I am still working on these. I kindly request some more time to complete this

**Comment by Rajgupta36** [comment]
> > @arkid15r Sir, I also have an idea of using manual API calls with debouncing to make the search more efficient. By implementing debouncing, we can reduce the number of API requests made while the user is typing, which would help avoid unnecessary calls and improve performance. This approach ensures that the search suggestions are updated in real-time without overwhelming the system with too many requests.
> 
> It seems your comment is more relevant to this issue -- #145. And this technique is already in place for the search component. However, I'm not sure if @Yashgupta9330 still works on the PR.

Yeah, I saw his PR; it includes debouncing. maybe @nitinawari approach is different

**Comment by arkid15r** [comment]
> @arkid15r Yes, I am still working on these. I kindly request some more time to complete this

I just want to make sure we are on the same page -- we've been just discussing the ways to tackle it. You didn't start any code yet, right? The search component PR itself is still [WIP](https://github.com/OWASP/Nest/pull/145). We want to replicate the [old frontend](https://nest.owasp.dev/projects/) behavior first and only after that start adding new functionality.

**Comment by nitinawari** [comment]
@arkid15r I haven’t started yet 😅. What do you suggest?

**Comment by arkid15r** [comment]
> @arkid15r I haven’t started yet 😅. What do you suggest?

No suggestions, just interested to see what would be your approach for tackling this.

**Comment by Rajgupta36** [comment]
@arkid15r ,  Algolia provides a client-side search bar component with autocomplete features.
 What are your thoughts on this?
https://www.algolia.com/doc/guides/solutions/ecommerce/search/autocomplete/federated-search/

**Comment by arkid15r** [comment]
> @arkid15r , Algolia provides a client-side search bar component with autocomplete features. What are your thoughts on this? https://www.algolia.com/doc/guides/solutions/ecommerce/search/autocomplete/federated-search/

This looks interesting. My idea was around [Algolia Query Suggestions](https://www.algolia.com/doc/guides/building-search-ui/ui-and-ux-patterns/query-suggestions/react/)

![Screenshot 2024-12-12 at 4 14 21 PM](https://github.com/user-attachments/assets/d0114c0f-d11f-4d2c-b8bd-1c937b84587a)

**Comment by Rajgupta36** [comment]
> > @arkid15r , Algolia provides a client-side search bar component with autocomplete features. What are your thoughts on this? https://www.algolia.com/doc/guides/solutions/ecommerce/search/autocomplete/federated-search/
> 
> This looks interesting. My idea was around [Algolia Query Suggestions](https://www.algolia.com/doc/guides/building-search-ui/ui-and-ux-patterns/query-suggestions/react/)
> 
> ![Screenshot 2024-12-12 at 4 14 21 PM](https://private-user-images.githubusercontent.com/2201626/395367815-d0114c0f-d11f-4d2c-b8bd-1c937b84587a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQxMTMyMjksIm5iZiI6MTczNDExMjkyOSwicGF0aCI6Ii8yMjAxNjI2LzM5NTM2NzgxNS1kMDExNGMwZi1kMTFmLTRkMmMtYjhiZC0xYzkzN2I4NDU4N2EucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTNUMTgwMjA5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZWJhZWU4MDhkNzk1MjE3YjVkZWFlYWVlMmE3YWE0M2JkYmYwMDgyYWYwYjBmZTgwZDYyZTVmYmVkODRiNjcwMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.YYf4tfgAVOAGBGmlH0pPbf944W1EA933jxbQ3tzcvj8)

that's good. i will try to work on it.

**Comment by Rajgupta36** [comment]
Hiii @arkid15r , can you assign this issue to me

**Comment by nitinawari** [comment]
@Rajgupta36  i am working on it

**Comment by Rajgupta36** [comment]
> @Rajgupta36  If you have a better approach, feel free to work on these as well. I am still figuring things out as I was focused on another issue.

👀

**Comment by nitinawari** [comment]
@Rajgupta36 you started working on this

**Comment by Rajgupta36** [comment]
> @Rajgupta36 you started working on this 

Yes. Also if we implement this we don't need reusable search bar. We can replace search bar with this component that algolia provide @arkid15r

**Comment by Rajgupta36** [comment]
> > @Rajgupta36 you started working on this 
> 
> Yes. Also if we implement this we don't need reusable search bar. We can replace search bar with this component that algolia provide @arkid15r

**Comment by nitinawari** [comment]
@Rajgupta36  ok complete this then, no problem

**Comment by Rajgupta36** [comment]
> @Rajgupta36  ok complete this then, no problem 

Thanks 👍

**Comment by arkid15r** [comment]
It seems our next step should be extending the search component w/ [autocomplete](https://www.algolia.com/doc/ui-libraries/autocomplete/introduction/getting-started/) feature

**Comment by yashpandey06** [comment]
@arkid15r if this is a free issue ...I can take this up ? ..let me know .

**Comment by arkid15r** [comment]
> @arkid15r if this is a free issue ...I can take this up ? ..let me know .
Last time I checked @Rajgupta36 was playing with Algolia index to implement it. 

@Rajgupta36 could you update this one?

**Comment by Rajgupta36** [comment]
okk i created a draft pr
